ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,641 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdint.h>
+#include <limits.h>
+#include <string.h>
+#include <dirent.h>
+#include <coll/rbt.h>
+#include <sys/queue.h>
+#include <time.h>
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "ldms_jobid.h"
+
+#define _GNU_SOURCE
+
+#define SAMP "lustre_client"
+#define LLITE_PATH "/proc/fs/lustre/llite"
+#define DEFAULT_SCHEMA_NAME "lustre_client"
+
+static ldmsd_msg_log_f log_fn;
+static char producer_name[LDMS_PRODUCER_NAME_MAX];
+static char schema_name[LDMS_SET_NAME_MAX];
+static uint64_t component_id;
+static time_t last_refresh;
+static time_t refresh_interval = 30;
+static int job_id_index;
+LJI_GLOBALS;
+
+/* red-black tree root for llites */
+static struct rbt llite_tree;
+
+struct llite_data {
+        char *fs_name;
+        char *name;
+        char *path;
+        char *stats_path;
+        ldms_set_t general_metric_set; /* a pointer */
+        struct rbn llite_tree_node;
+};
+
+static ldms_schema_t llite_general_schema;
+
+static char *llite_stats_uint64_t_entries[] = {
+        "dirty_pages_hits",
+        "dirty_pages_misses",
+        "read_bytes.sum",
+        "write_bytes.sum",
+        "brw_read.sum",
+        "brw_write.sum",
+        "ioctl",
+        "open",
+        "close",
+        "mmap",
+        "page_fault",
+        "page_mkwrite",
+        "seek",
+        "fsync",
+        "readdir",
+        "setattr",
+        "truncate",
+        "flock",
+        "getattr",
+        "create",
+        "link",
+        "unlink",
+        "symlink",
+        "mkdir",
+        "rmdir",
+        "mknod",
+        "rename",
+        "statfs",
+        "alloc_inode",
+        "setxattr",
+        "getxattr",
+        "getxattr_hits",
+        "listxattr",
+        "removexattr",
+        "inode_permission",
+        NULL
+};
+
+/*******************************************************************/
+/* ldms_metrict_by_name() employs a linear lookup and is verboten for
+ * use by plugins in the main sample() loop.  This section section
+ * implemants a faster lookup using the rbt structure, and offers
+ * alt_ldms_metric_by_name() as a limited alternative to the original.
+ * Someday, perhaps ldms_metric_by_name() will be improved and this
+ * section can be removed.
+ */
+/* red-black tree root for metric index lookup table */
+static struct rbt metric_lookup_tree;
```

Comment:
speed comment: use of rbt here is log(n_metrics) string compares, afaict. trading memory for speed with lib/src/coll/idx.h may be a better choice (o(1)). I can live with this version going to production and the improvement being done soonish.

Reply:
Fine.  I switched it to using coll/idx.h.  (would be nice to have some documentation in there)